### ISSUE TITLE
[bug] --build=missing does not honor CompatiblePackage

### ISSUE THREAD

**Issue by KerstinKeller** [conan-io/conan#6133]
### Environment Details (include every applicable attribute)
  * Operating System+version: Windows 10
  * Compiler+version: Visual Studio 2019
  * Conan version: 1.20.4
  * Python version: 3.8

### Steps to reproduce (Include if Applicable)
I have one recipe for `libA`, that defines a package compatibility range such as 

```
def package_id(self):
  if self.settings.compiler == "Visual Studio" and self.settings.compiler.version == "16":
    for version in ("15", "14"):
      compatible_pkg = CompatiblePackage(self)
      compatible_pkg.settings.compiler.version = version
      self.compatible_packages.append(compatible_pkg)
```

E.g. such that the package is binary downward compatible. If a consumer builds with VS2019, the package gets also accepted if it was build with VS2015 or VS2017.

This works fine if I have a consumer that uses libA (`self.requires.add("libA/0.1.1@user/stable")`) and do a:
```
conan install . 
```
It finds that package that was build with VS2015 and uses it.

However, if I specify 
```
conan install -b missing .
```
the package won't be accepted and Conan tries to rebuild `libA` with VS2019.
I would expect, that `libA` is not rebuild, because a matching package is available.

This behavior is especially "annoying", if the consumer requires a lot of packages, on of which are actually missing, but many of them are available as compatible packages Then all packages are rebuild instead of the single, actually missing package.

**Comment by memsharded** [comment]
Thanks for the report. Checked that the behavior is that one.

What I am currently struggling is to decide if it is a bug or a feature. The ``CompatiblePackage`` feature is provided as a **fallback**, its semantic is that the binary for your configuration is missing, but it can use another binary built with other configuration.

If we changed the behavior, and it didn't build from sources for the VS2019 compiler, then, suddenly there is no way to transitively build from sources those dependencies with the new VS2019 compiler version, which is a flow that makes a lot of sense. 

Maybe we need to think of other ways to enable both flows (maybe a ``--build=incompatible`` or the like), but so far we can't conclude this is a bug.

**Comment by KerstinKeller** [comment]
Yes, I figured that it might not be a bug.
I agree that it should be possible to force a rebuild of all packages which are not an exact match.
But at the same time it should also be possible to rebuild only actually missing packages, which is unfortunately not possible right now.

Maybe it is necessary to introduce another keyword.
But I guess then it would be more intuitive if `missing` only builds the actually missing packages, and `incompatible` also builds those which do not have the same package ID.
Then both could be combined to build both missing and incompatible packages.

However that would be breaking the current behavior of the `missing` keyword.
Also I am not sure if `incompatible` would be the right keyword, since these packages are compatible, but don't have the same package ID.

**Comment by memsharded** [comment]
I have done a small PoC, checking that it is not difficult to implement the behavior that we want: https://github.com/conan-io/conan/pull/6134

Yeah, naming is very difficult, after the PoC, I think that the new --build=keyword approach is the right one, but couldn't come with anything better yet, and what would be the default behavior. As this is only affected by the new ``compatible_package`` feature, which is experimental, I think it will not be breaking, whatever we decide the behavior is. Just lets try to define something that is reasonably understandable by users. Looping @lasote for a discussion when possible.

**Comment by KerstinKeller** [comment]
Ok, perfect, thanks for the quick reply 👍 
I am sure you will come up with something intuitive. 

As I said before: for me that would be `--build missing` -> build only missing packages, `--build compatible` -> build all packages which are compatible, but not an exact match.

A totally different approach would be to globally disable / enable compatibility mode.
E.g. you could say `CONAN_COMPATIBILITY_MODE=ON` or `CONAN_COMPATIBILITY_MODE=OFF`. For the first one, Conan accepts compatible packages and does not rebuild them when `--build missing` is specified. If the mode is switched to `OFF`, only exact matches are accepted and thus `--build missing` will rebuild all packages that are not binary compatible (current behavior). In this case you would not need to define an extra `--build` keyword.

I am not sure, which of the approaches fits Conan better.

**Comment by memsharded** [comment]
> As I said before: for me that would be --build missing -> build only missing packages, --build compatible -> build all packages which are compatible, but not an exact match.

Yeah, probably this is a good one. The other alternative, using global configuration, I think it is a bit more complex and difficult to understand, with the same effect, so I think I prefer the explicit ``--build`` one.

**Comment by memsharded** [comment]
This has been fixed, will be released soon in Conan 1.21
The behavior will be the intended: ``--build=missing`` will not build from sources if ``compatible_packages`` define that there is an existing compatible binary.

We have not implemented an alternative to build from sources if there are compatible binaries, until it is proved it is a requested feature and there is more evidence of the use case and user demand.
